### ISSUE TITLE
[IDEA] Automatically importing external web images through pasting the URL

### ISSUE THREAD

**Issue by EmpressFiona** [Jermolene/TiddlyWiki5#5286]
**Is your feature request related to a problem? Please describe.**
I'd like to be able to quickly paste external image URLs (or URLs in general, really) into TiddlyWiki as images. Doing so right now by pasting the image onto the page creates a text tiddler with the URL inside. The necessary functionality already exists; the 'img' syntax accepts external web images just as well as ones already on the wiki.

**Describe the solution you'd like**
Intelligent recognition of single incoming URLs pointing to recognized file types when pasting in, and turning them into a tiddler with either an image tag referencing the URL, or an encoded tiddler featuring the image downloaded from the URL.

Alternately, allowing the 'import attachment' button to work on URLs/URIs as well, or having a second button to import external files if you wish to preserve 'one-click' functionality, would work well enough too.

**Describe alternatives you've considered**
The alternative attempted is, of course, the creation of a tiddler which contains a reference to the URI. But that's not exactly a one-click process; while not that much slower, it's considerably slower than the 'speed of thought', so to speak, and seems to go against the design ethos of front-facing simplicity and back-end complexity.

An 'external URL' option on the 'img' syntax button on the toolbar would work, but again, not that much faster, albeit exposing the option easily to the end user.

**Additional context**
Adding images is already unintuitive because drag-and-drop isn't obvious, and the whole thing is mobile-unfriendly anyway (since drag and drop isn't really an option on mobile); in my opinion, the 'import attachment' button should be on the toolbar by default (and adding buttons to the toolbar is already something not easily noticed).

**Comment by EmpressFiona** [comment]
As an aside, pasting in an image copied to the clipboard creates a generic 'image.png' name regardless of whether there's already an 'image.png', and thus overwrites if you accept; it'd be easy enough to add a '2' or '3' to the end of that to be able to quickly create image albums with TiddlyWiki.

**Comment by Jermolene** [comment]
Thanks @EmpressFiona this is definitely an area where things could be improved.

Some of the challenges with all of this are the restrictions that browsers place on sandboxed web pages. Some of the constraints defy intuition, but have turned out to be important. For example, we can't in general retrieve the content from a URL, so we can't know whether https://example.com/thing points to an image or a web page. In some situations we may be able to use any extension portion of the URL but it's not a general solution.

I think the best we could do for pasting URLs might be to have extra keyboard shortcuts for "paste as image" or "paste as link".

It's a different story when TiddlyWiki is running under Node.js because these constraints don't exist. On paste, we'd have to transfer the URL to the server, and have the server retrieve the content and import it as a tiddler.

**Comment by saqimtiaz** [comment]
@Jermolene somewhat related, I quite like how Github handles pasting images into the editor, that is the image is saved in the background and a transclusion inserted in the editor at the caret.

I plan to look into something similar for TiddlyWiki, with an optional but enabled by default prompt asking the user if they wish to save the image. Combined with the ability to specify that the image be saved to the` /files` folder on Node.js, this would be a useful and quick workflow for adding images to a wiki.

**Comment by Jermolene** [comment]
> @Jermolene somewhat related, I quite like how Github handles pasting images into the editor, that is the image is saved in the background and a transclusion inserted in the editor at the caret.

Yes indeed, I think we can do that. It might still be reasonable workflow for @EmpressFiona (right click on image in browser and click "copy image", paste in TiddlyWiki editor)

> I plan to look into something similar for TiddlyWiki, with an optional but enabled by default prompt asking the user if they wish to save the image. Combined with the ability to specify that the image be saved to the` /files` folder on Node.js, this would be a useful and quick workflow for adding images to a wiki.

One frustration here is that we need to do as much as we can to make things smooth in the browser, and then inevitably have to build a very different pipeline for Node.js. Getting #3461 integrated might be a good start for opening up browser-node-browser transactions. There's a bunch of useful stuff like image compression we could do too.

**Comment by EmpressFiona** [comment]
> Some of the challenges with all of this are the restrictions that browsers place on sandboxed web pages. Some of the constraints defy intuition, but have turned out to be important. For example, we can't in general retrieve the content from a URL, so we can't know whether https://example.com/thing points to an image or a web page. In some situations we may be able to use any extension portion of the URL but it's not a general solution.

@Jermolene Wouldn't using the extension be simplest anyway? Most images on the web are going to have an extension if you're directly linking them. Obviously this isn't always the case, but it's the case often enough that you can make assumptions when the URL *does* have an extension, you can import it as an image.

@saqimtiaz's solution seems ideal, since you already have a confirmation dialogue for importing things anyway; you could even allow the user to specify if it's an image ahead of time if that can't be automatically determined. (You should probably allow a title as well, including trying to make one automatically from the posted URL, if you're keeping that step.)

Semantically, the way TiddlyWiki seems to want to work to the front-facing user is a common-sense approach. Pasting a link, particularly an *image* link, seems like the intent is more 'the content matters' than 'the text matters' - ie, you can quickly paste your content to a wiki. While the constraints you've mentioned of a non-node setup aren't something that allows for this to work universally, an approach can still likely be found that simplifies things for the user.

Honestly, browsing the issues to see if my problem was already brought up, it seems like the constraint of needing to support both the Node setup and the single-file setup is an immense headache for your codebase. Even getting the single-file wiki link has a gigantic disclaimer about how you need a method to save it, with node seeming to be the most popular and most flexible anyway...

**Comment by saqimtiaz** [comment]
@Jermolene I was thinking that on Node.js we have an extra optional setting to save an image being imported to an external file, and to have the file adaptor save the image to the files/ folder if for example, there is a `canonical_uri` field and the body has content, or a different field that serves a flag. I haven't had a chance to think this through in detail but I will delve into it post 5.1.23

Also of interest to @EmpressFiona might be the tweak I made [here](https://saqimtiaz.github.io/sq-tw/sandbox.html) which extends the insert image editor in the dropdown to have a droppable area where you can paste and import images and the transclusion is inserted into the editor.

Create a new tiddler and press <kbd>ctrl+shift+I</kbd>

![image](https://user-images.githubusercontent.com/68092/102255370-4ed90380-3f0a-11eb-95ce-b4232e83a380.png)

![image](https://user-images.githubusercontent.com/68092/102255506-78922a80-3f0a-11eb-8c42-6bc2a5b3c7d2.png)

**Comment by saqimtiaz** [comment]
> Even getting the single-file wiki link has a gigantic disclaimer about how you need a method to save it, with node seeming to be the most popular and most flexible anyway...

@EmpressFiona I am pretty sure that if we polled the community we would find that the single file users far outnumber those using Node.js

**Comment by joshuafontany** [comment]
Good thread. I ran into similar problems with my OEmbed plugin. This was my first attempt at a plugin that had different behavior if in a singke-file vs Node setup. It needed Bob's websocket setup to ask the server to call an Oembed API endpoint from a dictionary of public endpoints, and then saved the resulting embed json response with HTML code as a (system) tiddler, which until the 5.1.22 change would auto-sync back into the browser. My Iembed widget would pick up this change & refresh from a link to the target to transcluding the HTML from the response tiddler.

In single file wikis, it offered a prepared link to trigger the API call in a new tab & a textbox to paste the result. The user would have to do the "hard work" the browser could not of actually "fetching" the data.

I'll have to refactor for the change in how system tiddkers are synced now, but its work comparing the use-cases & methods.

https://github.com/joshuafontany/TW5-Oembed

**Comment by AnthonyMuscio** [comment]
On a related issue, on many social media sites pasting a URL will retrieve and display a preview, such as the featured image and a text excerpt. In many cases this may be enough for many users, and could even be very light if it is retrieved at load time.

Tones

**Comment by nathan-sain** [comment]
@saqimtiaz I use the node.js server and would love for an easy way to import images to as an external file. It's the only thing that really keeps me from using tiddlywiki. My use case is building role playing adventures and storing high resolution maps and background images is a requirement. I tried running a image archiving server called pinry and setting the _canonical_uri to my hosted images, but it became to much work to organize the images in both places. For now I use Joplin, but I love the database like features of tiddlywiki and would switch back in a heartbeat if I had a good workflow for using large images.

**Comment by saqimtiaz** [comment]
@nathan-sain Have a look at https://groups.google.com/g/tiddlywiki/c/T0MP1Adzzk0/m/RviYKuajBQAJ

**Comment by saqimtiaz** [comment]
> I quite like how Github handles pasting images into the editor, that is the image is saved in the background and a transclusion inserted in the editor at the caret.
> 
> I plan to look into something similar for TiddlyWiki, with an optional but enabled by default prompt asking the user if they wish to save the image. 

This has been implemented for TW 5.2.0 and can be tested in the [pre-release](https://tiddlywiki.com/prerelease/).

**Comment by twMat** [comment]
@saqimtiaz - a BIG thank you for making it possible to use the device screen clip tool to simply paste an image from memory directly into the editor - WHOA!!!! That is fantastic!

**Comment by CodaCodr** [comment]
@saqimtiaz shouldn't the Recent tab be updated, too?

**Comment by saqimtiaz** [comment]
> @saqimtiaz shouldn't the Recent tab be updated, too?

The default TW behaviour for importing images is to not give them a created or modified field. There have been no changes in this area. If you feel it should be changed, please create an issue.
